pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java

```diff
@@ -37,7 +38,7 @@ private boolean hasNoSubclasses(ASTClassOrInterfaceDeclaration klass) {
     }
 
     private boolean doesExtend(ASTAnyTypeDeclaration sub, ASTClassOrInterfaceDeclaration superClass) {
-        return sub != superClass && TypeTestUtil.isA(superClass.getTypeMirror(), sub);
+        return sub != superClass && TypeTestUtil.isA(superClass.getTypeMirror().getErasure(), sub);
```

Comment:
For now, I ignore the generic type here. This is most likely wrong, but I don't know where in the type system this problem is. For the test case now (see below) the rule seems to be working.

Reply:
I think your solution is fine. `Base<Object>` is definitely not a subclass of `Base<E>`, if it were we would have problems resolving types within the body of `Base`. It is however a parameterization of `Base`. Checking the erasure is the most correct way to check this I think.

Reply:
I thought, that we would replace the type var "E" by a concrete type at some point.... I was debugging the "capture" methods (e.g. the one called here https://github.com/pmd/pmd/blob/0fb4e08171cbc0c976244d88c26b6c0b5c04d552/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java#L407), but these seem to only replace wildcards. So `Base<E>` stayed `Base<E>` which is indeed something different than `Base<Object>`.

Reply:
Well here we only check whether the derived type has some parameterization of the superclass in its supertypes. But we don't need to know *which* parameterization, so using `isSubtype` is enough.

If you want to figure the parameterization of the superclass, you can use `getAsSuper`:
```java
class Base<E> {}
class Derived extends Base<Object> {}

JTypeMirror sub = Derived;
JTypeMirrot sup = Base<E>;

JTypeMirror asSuper = sub.getAsSuper(sup.getSymbol());
// asSuper == Base<Object>
```

This performs type variable replacement along the way. 

Throughout the entire type resolution framework type variables are treated nearly like real (concrete) types. They're not placeholder for other types, and don't just match "any type". This is the job of inference variables.